ArcGISRuntimeSDKQt_QMLSamples/Layers/ApplyMosaicRuleToRasters/ApplyMosaicRuleToRasters.qml

```diff
@@ -0,0 +1,113 @@
+// [WriteFile Name=ApplyMosaicRuleToRasters, Category=Layers]
+// [Legal]
+// Copyright 2020 Esri.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// [Legal]
+
+import QtQuick 2.6
+import QtQuick.Controls 2.2
+import Esri.ArcGISRuntime 100.10
+
+Rectangle {
+    id: rootRectangle
+    clip: true
+    width: 800
+    height: 600
+
+    property var rasterRules: ["None","NorthWest","Center","ByAttribute","LockRaster"]
+    property var mosaicRule: null
+
+    MapView {
+        id: mapView
+        anchors.fill: parent
+
+        Map {
+            BasemapTopographicVector {}
+
+            RasterLayer {
+                id: rasterLayer
+                ImageServiceRaster {
+                    id: imageServiceRaster
+                    url: "https://sampleserver7.arcgisonline.com/arcgis/rest/services/amberg_germany/ImageServer"
+                }
+                onLoadStatusChanged: {
+                    // Once loaded set the viewpoint to the center of the raster layer
+                    if (loadStatus === Enums.LoadStatusLoaded) {
+                        mapView.setViewpointCenterAndScale(rasterLayer.fullExtent.center, 25000);
+
+                        // Create a mosaic rule and set the mosaic rule property of the image service raster
+                        mosaicRule = ArcGISRuntimeEnvironment.createObject("MosaicRule");
```

Comment:
is it possible to create mosaic rule with declarative code, directly in ImageServiceRaster? In C++ code, the mosaic rule is set before the ImageServiceRaster is loaded.

Reply:
It's not a default property of `ImageServiceRaster` so I don't think so.

I don't think the order in which that first mosaic rule is set actually matters in relation to it being loaded. It's only initialized to a default `MosaicRule` because within the sample viewer it can pull some cached value for that which will alter which mosiac rule the sample starts with depending on where you left off.

Reply:
Sorry for delay to answer.
I'm not sure to understand why it's better to create mosaic rule like that, and not using declarative code. Something like:
```qml
ImageServiceRaster {
    id: imageServiceRaster
    url: "https://sampleserver7.arcgisonline.com/arcgis/rest/services/amberg_germany/ImageServer"
    mosaicRule: MosaicRule {
    }
}
```
                }

Reply:
@GuillaumeBelz Maybe the misunderstanding is on my part. Normally I don't attempt declarative code when it's not a default property of the QML class. When trying to do it declaratively one of two things happens, 

## First
```
ImageServiceRaster {
    id: imageServiceRaster
    url: "https://sampleserver7.arcgisonline.com/arcgis/rest/services/amberg_germany/ImageServer"
    mosaicRule: MosaicRule {
    }
}
```

this will result in an error `"ImageServiceRaster.mosaicRule" is not available in Esri.ArcGISRuntime 100.10` Which maybe it shouldn't.

## Second
```
ImageServiceRaster {
    id: imageServiceRaster
    url: "https://sampleserver7.arcgisonline.com/arcgis/rest/services/amberg_germany/ImageServer"
    MosaicRule {
    }
}
```

this had no tangible effect that I can see and if used as a replacement for line 49, it will not initialize the `MosaicRule` of the `ImageServiceRaster` which in the sample viewer will have some carryover effect when switching samples.

![carry-over-example](https://user-images.githubusercontent.com/11935261/97226250-f6e71180-1790-11eb-93ae-0a1f0e3653eb.gif)

Reply:
> Normally I don't attempt declarative code when it's not a default property of the QML class. When trying to do it declaratively one of two things happens,

To summary:

If a property is writable, it's possible to assign a value to this property in QML
```
MyComponent {
    myProperty: MyProperty {}
}
```

One property can be "default". In this case, this property is used when the name of the property is not specified in QML. The type of the object must be compatible with the type of the property.
```
// Q_CLASSINFO("DefaultProperty", "myProperty") in C++ class definition

MyComponent {
    MyProperty {} // same as "myProperty: MyProperty {}"
}
```

A special case of default property is `objects`, which take a list of QmlObject. In this case, all the children which are not assigned to a specific property are added to the `objects` property. The setter of `objects` checks the type of the objects and assigns them to the property correspondent to the type of the object.

```
// Q_CLASSINFO("DefaultProperty", "object") in C++ class definition

MyComponent {
    MyProperty {} // assigned to myProperty
    Rectangle {} // assigned to the property which accept Rectangle
}
```

If the child object is not assigned to a property and the setter can't find the correct property, nothing happens.

> # First
> this will result in an error "ImageServiceRaster.mosaicRule" is not available in Esri.ArcGISRuntime 100.10 Which maybe it shouldn't.

`mosaicRule` is writable in `QmlImageServiceRaster` (since 100.9) then the code should be valid. I don't know what is the issue, but this is a bug.

> #Second
> this had no tangible effect that I can see and if used as a replacement for line 49, it will not initialize the MosaicRule of the ImageServiceRaster which in the sample viewer will have some carryover effect when switching samples.

`ImageServiceRaster` have a `objects` default property and `mosaicRule` is writable, then this code should work too. I checked the code of `ImageServiceRaster`, this feature is not supported yet. We should add that.

We can create 2 issues, to fix the bug with `mosaicRule` property and to add the support of `MosaicRule` in `objects` property

 @JamesMBallard @ldanzinger

Reply:
>We can create 2 issues, to fix the bug with mosaicRule property and to add the support of MosaicRule in objects property

@jared-2016 @GuillaumeBelz please work together to get those logged.

One is a trivial fix and the other (add default property support) will take a bit more effort.

Reply:
Thank you for the added explanation! I understand the problem now. I can log both of the issues.